docs/source/torchao_vllm_integration.md

```diff
@@ -0,0 +1,398 @@
+
+To make a new TorchAO quantization method work with VLLM, you need to implement minimal tensor subclass operations that support **tensor parallelism**. VLLM uses `narrow()` and `copy_()` to move data from host cpu loaded in a state dict to the device, these  require these specific aten operations:
+
+(why-these-operations)=
+### Why these ?
+
+VLLM's tensor parallelism works by:
+1. **{meth}`~torch.Tensor.narrow`** - Slicing weight tensors across different dimensions
+2. **Sharding** - Distributing tensor chunks across multiple GPUs
+3. **{meth}`~torch.Tensor.copy_`** - Moving tensor data between devices
+4. **{meth}`~torch.Tensor.detach`**
+
+
+A helpful pattern for doing this is `_apply_fn_to_data`, a method that applies a given function to all the attributes on your class w/ Tensor types. Below is a generic implementation that should work for most subclasses. We make heavy use of this, you can see an examples in {doc}`mx_ops.py <../torchao/prototype/mx_formats/mx_ops.py>`:
+
+```python
+def _apply_fn_to_data(self, fn: Callable):
+    """Applies a fn to all tensor components stored on this class"""
+    tensor_names, ctx = self.__tensor_flatten__()
+
+    # Apply the function to each tensor component
+    new_tensors = {}
+    for name in tensor_names:
+        new_tensors[name] = fn(getattr(self, name))
+
+    return self.__class__.__tensor_unflatten__(
+        new_tensors,
+        ctx,
+        None,  # outer_size parameter
+        None,  # outer_stride parameter
+    )
+```
+
+(step-by-step-guide)=
+## Step-by-Step Guide to Add a New Quantization Method
+
+(create-tensor-subclass)=
+#### 1. Create Your Tensor Subclass
+
+```python
+from torchao.core.config import AOBaseConfig
+from torchao.utils import TorchAOBaseTensor
+
+@dataclass
+class MyNewQuantConfig(AOBaseConfig):
+    """Configuration for your new quantization method"""
+    bits: int = 8
+    VERSION: ClassVar[int] = 1
+
+class MyQuantizedTensor(TorchAOBaseTensor):
+    """Example based on FbgemmFp8Tensor - stores quantized data + scale"""
+
+    tensor_data_attrs = ["quantized_data", "scale"]
+    tensor_attributes = ["dtype"]
+
+    def __new__(cls, quantized_data, scale, dtype):
+        shape = quantized_data.shape
+        return torch.Tensor._make_wrapper_subclass(
+            cls, shape, device=quantized_data.device, dtype=dtype, requires_grad=False
+        )
+
+    def __init__(self, quantized_data, scale, dtype):
+        self.quantized_data = quantized_data
+        self.scale = scale
+
+    def __tensor_flatten__(self) -> Tuple[List[str], List]:
+        """Serialize tensor subclass into plain tensors and metadata"""
+        return self.tensor_data_attrs, [getattr(self, attr) for attr in self.tensor_attributes]
+
+    @classmethod
+    def __tensor_unflatten__(cls, tensor_data_dict: Dict[str, torch.Tensor], tensor_attributes: List,
+                           outer_size: Optional[torch.Size], outer_stride: Optional[Tuple]) -> "MyQuantizedTensor":
```

Comment:
formatting is weird here, maybe just do 1 per line?
<img width="666" alt="Screenshot 2025-06-11 at 7 05 07 PM" src="https://github.com/user-attachments/assets/2d86d1c3-0d12-4eaf-8dbc-7963c205d4af" />